### ISSUE TITLE
QueryBuilder to remember last query

### ISSUE THREAD

**Issue by mfrasca** [Ghini/ghini.desktop#132]
_From @mfrasca on December 24, 2015 15:48_

in a comment to #36, @RoDuth writes:

> In general I think the search functions are one of the big strengths of Bauble and the query builder is great for a layman like myself. I always thought that it would be good if the query builder could stay where you last had it each time you open it and have a clear button to reset it. So that you could use it to generate a search then open it back up and its in the same place you left it so you can add extra criteria to "narrow the search" so to speak.

my behaviour relating to the above is to have the QB construct a query then I edit it by hand and do not go back to the QB unless I need a completely different search.

_Copied from original issue: Bauble/bauble.classic#238_

**Comment by mfrasca** [comment]
_From @RoDuth on December 25, 2015 6:37_

I agree, I also tend to edit by hand but I have a couple of scenarios where returning to the previous state would be handy.  Assuming its no harder than clicking the reset button to clear it anyway.
* When I'm trying to find the best approach to get the results I'm looking for I often use the query builder to "fish" for domains/expressions/properties that work.  It would be good to open back up to where I last was so I can "fish" again from the same place.  There are lots of properties that I don't remember unless I use them often.
* When using the query builder to design long, ugly, complex search strings with lots of ANDs or ORs (e.g. `accession where plants.location.code like "01.0%" and source.source_detail.name like "SGAP%" or source.source_detail.name like "Irene%" and plants.quantity > "0"`) I often will "build" them in stages, checking at each stage that I am getting closer to what I am looking for.  Nothing more frustrating than entering one of these long strings and getting a "couldn't find anything for.." result and then having to work out which bit went wrong.
* To narrow the result by adding one more expression. (e.g. in the above example it hadn't occurred to me that I would also get dead accessions and so I had to add the `and plants.quantity > "0"`, maybe a bad example as this would be a simple hand edit but you get the point)  I think this narrowing of the search is the main benefit really.  The others are kind of related.

**Comment by mfrasca** [comment]
Maybe a note as to explain why this is taking me so much time. I would not want to just "remember the last query", I would want to parse the query string as it is in the `comboboxentry`, the place where you type the query, and where the `QueryBuilder` puts it.
to parse, in information science, means to read, interpret, recognize the structure, build a model for it — the `SearchParser` already does this when we prepare to execute the query, was it typed by hand or generated by the `QueryBuilder`.
had the `QueryBuilder` been designed according to the Model-View-Presenter pattern, once we have the parsed query, we have the model we want to edit, we just handle it to the `QueryBuilder` (a Presenter), it would initialize the view with it, and we would be done.
unfortunately, the `QueryBuilder` is just a very smart `gtk.Dialog`, it implements logic, owns the model, computes its string representation, all in one. And the model is completely unrelated to the `pyparsing` model we use to execute queries.
bottom line?
I would want to completely rewrite the `QueryBuilder`, as I have rewritten the `SearchParser`.

**Comment by mfrasca** [comment]
it works but there's a small detail which makes is still suboptimal: 
see the difference when building the query »``accession where prov_type = 'Purchase'``«, or when opening the query builder starting from the same query.
![screenshot from 2017-06-30 10-40-43](https://user-images.githubusercontent.com/122869/27743187-c9f40a66-5d80-11e7-9c9e-352dab9b9710.png)
![screenshot from 2017-06-30 10-40-50](https://user-images.githubusercontent.com/122869/27743188-c9f8d140-5d80-11e7-9c49-da1a08fe2bf2.png)

**Comment by mfrasca** [comment]
I'm working at it further, not so much more code, and you will get back to the same place.
